packages/cardano-services/src/Program/utils.ts

```diff
@@ -1,58 +1,89 @@
+import { ProgramOptionDescriptions } from './ProgramOptionDescriptions';
+import { ProviderError, ProviderFailure, TxSubmitProvider } from '@cardano-sdk/core';
+import { RabbitMqTxSubmitProvider } from '@cardano-sdk/rabbitmq';
+import { ServiceNames } from './ServiceNames';
+import { WebSocketClosed, ogmiosTxSubmitProvider, urlToConnectionConfig } from '@cardano-sdk/ogmios';
 import Logger from 'bunyan';
-import dns from 'dns';
+import dns, { SrvRecord } from 'dns';
 import pRetry, { FailedAttemptError } from 'p-retry';
 
-export const RETRY_BACKOFF_FACTOR_DEFAULT = 1.1;
-export const RETRY_BACKOFF_MAX_TIMEOUT_DEFAULT = 60 * 1000;
+export const SERVICE_DISCOVERY_BACKOFF_FACTOR_DEFAULT = 1.1;
+export const SERVICE_DISCOVERY_BACKOFF_TIMEOUT_DEFAULT = 60 * 1000;
+export const DNS_SRV_CACHE_KEY = 'DNS_SRV_Record';
 
-type RetryBackoffConfig = {
-  factor: number;
-  maxRetryTime: number;
+export type RetryBackoffConfig = {
+  factor?: number;
+  maxRetryTime?: number;
 };
 
 export const onFailedAttemptFor =
-  (operation: string, logger: Logger) =>
+  (serviceName: string, logger: Logger) =>
   async ({ attemptNumber, message, retriesLeft }: FailedAttemptError) => {
     const nextAction = retriesLeft > 0 ? 'retrying...' : 'exiting';
     logger.trace(message);
-    logger.debug(`${operation}: Attempt ${attemptNumber} of ${attemptNumber + retriesLeft}, ${nextAction}`);
+    logger.debug(
+      `Establishing connection to ${serviceName}: Attempt ${attemptNumber} of ${
+        attemptNumber + retriesLeft
+      }, ${nextAction}`
+    );
     if (retriesLeft === 0) {
       logger.error(message);
-      // await server.shutdown(); ?
-      // eslint-disable-next-line unicorn/no-process-exit
-      process.exit(1);
+      // Invokes onDeath() callback within cardano-services entrypoints, following by server.shutdown() and process.exit(1)
+      process.kill(process.pid, 'SIGTERM');
     }
   };
 
-export const resolveDnsSrvWithExponentialBackoff = async (
-  serviceName: string,
-  { factor, maxRetryTime }: RetryBackoffConfig,
-  logger: Logger
-) =>
-  await pRetry(
-    async () => {
-      // Shall we grab the first one always?
-      const [record] = await dns.promises.resolveSrv(serviceName);
-      return record;
-    },
-    {
-      factor,
-      maxRetryTime,
-      onFailedAttempt: onFailedAttemptFor(`Establishing connection to ${serviceName}`, logger)
-    }
-  );
+// Get a random selection of dns srv resolved service address initially and make it sticky for reconnects by storing a reference in memory
+export const getStikyAddressWithDnsSrv = async (serviceName: string, cache: InMemoryCache): Promise<SrvRecord> => {
+  const stickyAddress = cache.getVal<SrvRecord>(`${DNS_SRV_CACHE_KEY}/${serviceName}`);
+  if (!stickyAddress) {
+    const [address] = await dns.promises.resolveSrv(serviceName);
+    cache.set(`${DNS_SRV_CACHE_KEY}/${serviceName}`, address, UNLIMITED_CACHE_TTL);
```

Comment:
As I see from the commit it will default to `#ttlDefault` which is [120 mins](https://github.com/input-output-hk/cardano-js-sdk/blob/a2bfcc62c25e71d78d07b961267d7ce9679b6cf4/packages/cardano-services/src/InMemoryCache/defaults.ts#L1) but in this case, here we would like to cache it during the whole `cardano-services` instance lifetime, no? @rhyslbw

Reply:
Ah yes, sorry, I thought it was going to use the underlying library default